client/worker.go

```diff
@@ -0,0 +1,241 @@
+	headerRetryAfter                  = "Retry-After"
+	headerContentEncoding             = "Content-Encoding"
+	headerContentType                 = "Content-Type"
+	headerValueGZIP                   = "gzip"
+	headerValueXProtobuf              = "application/x-protobuf"
+)
+
+type worker struct {
+	authToken  string
+	client     *http.Client
+	endpoint   string
+	gzipWriter *gzip.Writer
+	maxRetries uint64
+	doneCh     chan struct{}
+}
+
+func newWorker(client *http.Client, endpoint string, authToken string, maxRetries uint64) *worker {
+	w := &worker{
+		authToken:  authToken,
+		client:     client,
+		endpoint:   endpoint,
+		gzipWriter: gzip.NewWriter(nil),
+		maxRetries: maxRetries,
+		doneCh:     make(chan struct{}),
+	}
+	return w
+}
+
+func (w *worker) export(ctx context.Context, batch *jaegerpb.Batch) error {
+	_, span := trace.StartSpan(ctx, "export")
+	defer span.End()
+
+	if len(batch.Spans) == 0 {
+		span.AddAttributes(trace.BoolAttribute("empty_batch", true))
+		return nil
+	}
+
+	sr, err := w.prepare(ctx, batch)
+	if err != nil {
+		recordEncodingFailure(ctx, sr)
+		span.SetStatus(trace.Status{
+			Code: trace.StatusCodeInvalidArgument,
+		})
+		return err
+	}
+
+	retries := uint64(0)
+
+	for {
+		err := w.send(ctx, sr)
+		if err == nil {
+			recordSuccess(ctx, sr)
+			break
+		}
+		recordSendFailure(ctx, sr)
+
+		if err.Permanent {
+			span.SetStatus(trace.Status{
+				Code: ocTrancelator.OCStatusCodeFromHTTP(int32(err.StatusCode)),
+			})
+			recordDrops(ctx, sr)
+			return err
+		}
+
+		if err.RetryDelaySeconds > 0 {
+			if retries >= w.maxRetries {
+				span.SetStatus(trace.Status{
+					Code: trace.StatusCodeDeadlineExceeded,
+				})
+				recordDrops(ctx, sr)
+				err.Permanent = true
+				return err
+			}
+			time.Sleep(time.Duration(err.RetryDelaySeconds) * time.Second)
+			retries++
+			continue
+		}
+		return err
+	}
+	return nil
+}
+
+func (w *worker) send(ctx context.Context, r *sendRequest) *ErrHTTPSend {
+	_, span := trace.StartSpan(ctx, "export")
+	defer span.End()
+
+	req, err := http.NewRequest("POST", w.endpoint, bytes.NewBuffer(r.message))
+	if err != nil {
+		span.SetStatus(trace.Status{
+			Code:    trace.StatusCodeInvalidArgument,
+			Message: err.Error(),
+		})
+		return &ErrHTTPSend{Err: err, Permanent: true}
+	}
+	req.Header.Add(headerContentType, headerValueXProtobuf)
+	req.Header.Add(headerContentEncoding, headerValueGZIP)
+	if w.authToken != "" {
+		req.Header.Add(headerAuthToken, w.authToken)
+	}
+
+	resp, err := w.client.Do(req)
+	if err != nil {
+		span.SetStatus(trace.Status{
+			Code:    trace.StatusCodeInternal,
+			Message: err.Error(),
+		})
+		return &ErrHTTPSend{Err: err}
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode >= 200 && resp.StatusCode <= 299 {
+		return nil
+	}
+
+	// Drop the batch if server thinks it is malformed in some way or client is not authorized
+	if resp.StatusCode == http.StatusBadRequest || resp.StatusCode == http.StatusUnauthorized {
```

Comment:
Question: should we drop on Unauthorized or log and backoff continuously? The later would allow the operator of the Collector to notice the problem and fix it, hopefully before the data is lost. What do you think? Or is it too optimistic and we are just going to needlessly accumulate data in the queues?

Reply:
I think we should drop and record metric. Even if we don't drop and keep it in mem for some time so someone can take a look, as it stands right now, the person will have to change config and restart the collector to fix it. So it'll be lost anyway. May be we can revisit if/when Otel supports hot-starts on config change?

Reply:
Good point.